reader_concurrency_semaphore.cc

```diff
@@ -1274,8 +1274,8 @@ bool reader_concurrency_semaphore::has_available_units(const resources& r) const
     return (_resources.non_zero() && _resources.count >= r.count && _resources.memory >= r.memory) || _resources.count == _initial_resources.count;
 }
 
-bool reader_concurrency_semaphore::all_need_cpu_permits_are_awaiting() const {
-    return _stats.need_cpu_permits == _stats.awaits_permits;
+bool reader_concurrency_semaphore::cpu_concurrency_limit_reached() const {
+    return (_stats.need_cpu_permits - _stats.awaits_permits) >= _cpu_concurrency();
```

Comment:
this might run into unsigned integer underflow, see https://godbolt.org/z/TaxMYxMjz

so probably we should put something like:

```c++
uint64_t provisioned = _stats.awaits_permits + _cpu_concurrency();
return _stats.need_cpu_permits >= provisioned;
```

Reply:
This suggests that there is a bug somewhere: all permits that are in awaits state, are also accounted in the `need_cpu_permits` state too.
See https://github.com/scylladb/scylladb/blob/591876b44e5a38052c9ab2d9d4f219bcf87275a9/reader_concurrency_semaphore.cc#L1543-L1547

Reply:
No, because `await` is a sub-state of `need_cpu`, `_stats.need_cpu_permits >= _stats.awaits_permits` is an invariant, protected by asserts.

Reply:
My comment raced with Botond's, it was directed at @tchaikov.

Reply:
Thanks guys. I see, so it's a red herring. I am chasing a bug in this area. And thus caught my attention.